[SPARK-24283][ML] Make ml.StandardScaler skip conversion of Spar…

sujithjay opened the pull request:
…k ml vectors to mllib vectors

## What changes were proposed in this pull request?
Currently, ml.StandardScaler converts Spark  ml vectors to mllib vectors during transform operation. This PR makes changes to skip this step.

## How was this patch tested?
Existing tests in StandardScalerSuite

Commit:
[SPARK-24283][ML][WIP] Make ml.StandardScaler skip conversion of Spark ml vectors to mllib vectors

holdensmagicalunicorn commented:
@sujithjay, thanks! I am a bot who has found some folks who might be able to help with the review:@mengxr, @jkbradley and @MLnick

AmplabJenkins commented:
Can one of the admins verify this patch?

sujithjay commented:
@holdenk Could you also please take a look at this PR?

AmplabJenkins commented:
Can one of the admins verify this patch?

AmplabJenkins commented:
Can one of the admins verify this patch?

hhbyyh commented on mllib/src/main/scala/org/apache/spark/ml/feature/StandardScaler.scala:
```diff
@@ -160,15 +160,89 @@ class StandardScalerModel private[ml] (
+    *         for the column with zero std.
+    */
+  @Since("2.3.0")
+  def transform(vector: Vector): Vector = {
```
private[spark]?

hhbyyh reviewed (commented):
IMO it's a good thing that we move the implementation from MLlib to ML. Yet I do see one concern about test coverage. Most of the unit tests are still in MLlib and ML does not provide equivalent converge, which makes it difficult to verify code changes. 

One solution is to move all the unit tests to ML too, though this will double the unit test time.

Commit:
[SPARK-24283][ML] Change access modifier of `transform` method

Change access modifier of `transform` method from public to
private[spark]. Also, remove `@Since` annotation from the method.

sujithjay commented:
@hhbyyh I, personally, like the idea of moving the unit tests to ML. However, are you suggesting I move some of them as part of this PR? Or were you suggesting that it is a task we need to take up separately?

hhbyyh commented:
I think it's better to move the code and unit test in one PR. But since it's not a trivial change, I suggest you to wait for committers' opinion.

sujithjay commented:
OK to test

sujithjay commented:
Hi @mengxr, @jkbradley, @MLnick and @holdenk , could you please review this PR ? Thank you.

sujithjay commented:
Hi @mengxr, @jkbradley, @MLnick, @holdenk, @viirya , could you please review this PR ?

holdenk commented:
Sorry I've been slow to respond, I'll take a look today.

holdenk commented on mllib/src/main/scala/org/apache/spark/ml/feature/StandardScaler.scala:
```diff
@@ -160,15 +160,88 @@ class StandardScalerModel private[ml] (
+   * `lazy val` so it will be evaluated in that branch. Note that we don't
+   * want to create this array multiple times in `transform` function.
+   */
+  private lazy val shift: Array[Double] = mean.toArray
```
How does this interplay with serialization? Would it make sense to evaluate this before we serialize the UDF so it isn't done on each worker or?

holdenk commented on mllib/src/main/scala/org/apache/spark/ml/feature/StandardScaler.scala:
```diff
@@ -160,15 +160,88 @@ class StandardScalerModel private[ml] (
+        }
+      }
+      Vectors.dense(values)
+    } else if ($(withStd)) {
```
Maybe leave a comment withStd and not mean since when tracing the code by hand the nested if/else if can get a bit confusing flow wise.

holdenk reviewed (commented):
Thanks for working to get us off of MLlib :) I'm +1 on the need to move the tests over, it will temporarily increase our total test time but that's better than not having coverage. If you wouldn't mind a quick micro-benchmark showing that it's not a performance regression over the old code-path would be great as well.

mgaido91 commented:
@sujithjay are you still active on this PR? Or shall someone take this over?

srowen reviewed (commented):
@zhengruifeng are you working on something similar or would you care to review?

zhengruifeng commented:
@srowen  After the crrently merged pr [Make the transform natively in ml framework to avoid extra conversion](https://github.com/apache/spark/pull/24963), `StandardScaler` can now skip the conversion .ml.vector => .mllib.vector

And in [RobustScaler](https://github.com/apache/spark/pull/25160), I am to make the transformation more common, so that `RobustScaler` can reuse the function.
